internals/plan/plan.go

```diff
@@ -641,179 +646,218 @@ func CombineLayers(layers ...*Layer) (*Layer, error) {
 		}
 	}
 
-	// Ensure fields in combined layers validate correctly (and set defaults).
-	for name, service := range combined.Services {
-		if service.Command == "" {
-			return nil, &FormatError{
-				Message: fmt.Sprintf(`plan must define "command" for service %q`, name),
-			}
+	// Set defaults where required.
+	for _, service := range combined.Services {
+		if !service.BackoffDelay.IsSet {
+			service.BackoffDelay.Value = defaultBackoffDelay
+		}
+		if !service.BackoffFactor.IsSet {
+			service.BackoffFactor.Value = defaultBackoffFactor
 		}
+		if !service.BackoffLimit.IsSet {
+			service.BackoffLimit.Value = defaultBackoffLimit
+		}
+	}
+
+	for _, check := range combined.Checks {
+		if !check.Period.IsSet {
+			check.Period.Value = defaultCheckPeriod
+		}
+		if !check.Timeout.IsSet {
+			check.Timeout.Value = defaultCheckTimeout
+		}
+		if check.Threshold == 0 {
+			// Default number of failures in a row before check triggers
+			// action, default is >1 to avoid flapping due to glitches. For
+			// what it's worth, Kubernetes probes uses a default of 3 too.
+			check.Threshold = defaultCheckThreshold
+		}
+	}
+
+	return combined, nil
+}
+
+// Validate checks that the layer is valid.
+// See also Plan.Validate, which does additional checks based on the combined
+// layers.
+func (l *Layer) Validate() error {
```

Comment:
Can we change this receiver name from `l` to `layer` to avoid it being confused with 1 or I? Receiver names are normally 1-letter in Go, but that's not a hard and fast rule if there are reasons not to.